File: app/src/main/java/com/example/workoutpixel/Database/AppDatabase.java
package com.example.workoutpixel.Database;

import android.content.Context;

import androidx.room.Database;
import androidx.room.Room;
import androidx.room.RoomDatabase;

import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

@Database(entities = {PastWorkout.class, Widget.class}, version = 1, exportSchema = false)
public abstract class AppDatabase extends RoomDatabase {
    private static final int NUMBER_OF_THREADS = 4;
    static final ExecutorService databaseWriteExecutor =
            Executors.newFixedThreadPool(NUMBER_OF_THREADS);
    private static final Object sLock = new Object();
    private static AppDatabase pastWorkoutsDb;

    // Return database - if there already exists one, return this one, otherwise create one.
    public static AppDatabase getDatabase(Context context) {
        synchronized (sLock) {
            if (null == pastWorkoutsDb) {
                pastWorkoutsDb = buildDatabaseInstance(context);
            }
            return pastWorkoutsDb;
        }
    }

    private static AppDatabase buildDatabaseInstance(Context context) {
        return Room.databaseBuilder(context,
                AppDatabase.class,
                "WorkoutPixelDatabase")
                .allowMainThreadQueries()
                .build();
    }

    public abstract WorkoutDao workoutDao();

    public void cleanUp() {
        pastWorkoutsDb = null;
    }

}

File: app/src/main/java/ch/karimattia/workoutpixel/data/AppDatabase.kt
package ch.karimattia.workoutpixel.data

import android.content.Context
import androidx.room.*
import androidx.room.migration.AutoMigrationSpec
import dagger.Module
import dagger.Provides
import dagger.hilt.InstallIn
import dagger.hilt.android.qualifiers.ApplicationContext
import dagger.hilt.components.SingletonComponent
import javax.inject.Singleton


@Database(
	entities = [GoalWithoutCount::class, PastClick::class],
	views = [Goal::class],
	version = 9,
	autoMigrations = [
		AutoMigration(from = 2, to = 4, spec = AppDatabase.AutoMigration::class),
		AutoMigration(from = 3, to = 4, spec = AppDatabase.AutoMigration::class),
		AutoMigration(from = 4, to = 5, spec = AppDatabase.AutoMigration::class),
		AutoMigration(from = 5, to = 6, spec = AppDatabase.AutoMigration::class),
		AutoMigration(from = 6, to = 7, spec = AppDatabase.AutoMigration::class),
		AutoMigration(from = 7, to = 8, spec = AppDatabase.AutoMigration::class),
		AutoMigration(from = 8, to = 9, spec = AppDatabase.AutoMigration::class),
	])
abstract class AppDatabase : RoomDatabase() {
	abstract fun goalDao(): GoalDao

	@DeleteColumn(tableName = "goals", columnName = "status")
	class AutoMigration : AutoMigrationSpec
}

@InstallIn(SingletonComponent::class)
@Module
class DatabaseModule {

	@Provides
	fun provideGoalDao(appDatabaseKotlin: AppDatabase): GoalDao {
		return appDatabaseKotlin.goalDao()
	}

	@Provides
	@Singleton
	fun provideAppDatabase(@ApplicationContext context: Context): AppDatabase {
		return Room.databaseBuilder(
			context,
			AppDatabase::class.java,
			"WorkoutPixelDatabase"
		)
			//.allowMainThreadQueries()
			//.addMigrations(MIGRATION_1_2)
			//.fallbackToDestructiveMigration()
			.build()
	}
}
